src/logic/function/basic.lean

```diff
@@ -540,14 +540,23 @@ along a function `f : α → β` to a function `β → γ`,
 by using the values of `g` on the range of `f`
 and the values of an auxiliary function `e' : β → γ` elsewhere.
 
-Mostly useful when `f` is injective. -/
+Mostly useful when `f` is injective,
+more generally when `f a  f b → g a = g b`. -/
 def extend (f : α → β) (g : α → γ) (e' : β → γ) : β → γ :=
 λ b, if h : ∃ a, f a = b then g (classical.some h) else e' b
 
 lemma extend_def (f : α → β) (g : α → γ) (e' : β → γ) (b : β) [decidable (∃ a, f a = b)] :
   extend f g e' b = if h : ∃ a, f a = b then g (classical.some h) else e' b :=
 by { unfold extend, congr }
 
+lemma extend_apply_of_unique (g : α → γ) (e' : β → γ)
```

Comment:
`g` can be implicit here, since it appears in `hf`, but this is up to you.

Similarly below.

Reply:
I have no opinion, I was willing to stay as close as possible to the former syntax.
(Also, in the hypothesis `hf`, I wanted to have `a` and `b` optional, but then this breaks applications of `exact hf` because at that point, one needs to `intro a b`…)

Reply:
Is `(hf : ∀ ⦃a, b⦄, f a = f b → g a = g b)` what you want? You can read about `⦃...⦄` in section 6.5 of TPIL.

Reply:
Cool, that works ! I just uploaded a revision.